tools/concepts/mfcc/test_mfcc.m

```diff
@@ -0,0 +1,219 @@
+			i4 = param.fft_size;
+			i5 = (i - 1) * param.hop_size + 1;
+			i6 = i5 + param.hop_size - 1;
+			fft_buf(i3:i4) = state.buffer(i5:i6);
+		        %figure(1); plot(fft_buf); grid on;
+
+			% FFT -> power spectra
+			lin_p = do_win_fft(state, fft_buf);
+			%figure(2); plot(lin_p); grid on;
+
+			% Mel bands -> dB -> DCT
+			mel_db = convert_lin_to_mel_db(state, param, lin_p);
+			%figure(3); plot(mel_db); grid on;
+			mel_db_matrix(:,i) = mel_db;
+			cc = dct(mel_db); % DCT type-II (default)
+			%cc = dct(mel_db, 'Type', 3); % DCT type-III
+			cc_matrix(:,i) = cc;
+
+			% Update prev
+			state.prev = fft_buf(param.hop_size + 1:end);
+		end
+
+		% Move unused samples to beginning of buffer, update position
+		remain = i2 - m * param.hop_size;
+		if remain > 0
+			state.buffer(1:remain) = state.buffer(i6 + 1:i6 + remain);
+			state.buffer_pos = remain + 1;
+		else
+			state.buffer_pos = 1;
+		end
+	end
+
+end
+
+function mel_db = convert_lin_to_mel_db(state, param, lin_p)
+
+	mel_db = zeros(param.mel_bands, 1);
+	for i = 1:param.mel_bands
+		mel_db(i) = 10*log10(sum(lin_p .* state.triangles(:, i)));
+	end
+
+end
+
+function state = init_mfcc(param, stream)
+
+	state.prev_size = param.fft_size - param.hop_size;
+	state.prev = zeros(state.prev_size, 1);
+	state.buffer_size = param.fft_size - state.prev_size + stream.frames;
+	state.buffer = zeros(state.buffer_size, 1);
+	state.buffer_pos = 1;
+
+	% TODO: Add needed windows, hanning is just a guess to get started
+	switch lower(param.window)
+		case 'hanning'
+			state.win = hanning(param.fft_size, 'periodic');
+		otherwise
+			error('Non-supported window');
+	end
+
+	% FFT frequencies
+	fn = stream.fs / 2;
+	state.half_fft_size = param.fft_size / 2 + 1;
+	state.fft_f = linspace(0, fn, state.half_fft_size);
+
+	% Mel frequencies
+	ms = hz_to_mel(0);
+	me = hz_to_mel(fn);
+	mels = linspace(ms, me, param.mel_bands + 2);
+	state.mel_band_hz = mel_to_hz(mels);
+	state.mel_idx = 1 + floor((state.half_fft_size - 1) * state.mel_band_hz / fn);
+
+	% Lin Hz to Mel triangles
+	% Note: In firmware it's a good idea to pack the triangles to a single vector
+	state.triangles = zeros(state.half_fft_size, param.mel_bands);
+	for i = 1:param.mel_bands
+		i1 = state.mel_idx(i);
+		i2 = state.mel_idx(i+1);
+		i3 = state.mel_idx(i+2);
+		s = 1 / (i3 - i1);
+		% Scale tallest triangle peak to 1.0
+		if i == 1
+			c =  1 / s;
+		end
+		state.triangles(i1:i2, i) = linspace(0, 1, i2 - i1 + 1) * s * c;
+		state.triangles(i2:i3, i) = linspace(1, 0, i3 - i2 + 1) * s * c;
+	end
+	if 1
+		figure;
+		plot(state.fft_f, state.triangles);
+		grid on;
+		xlabel('Hz');
+		ylabel('Weight')
+	end
+
+end
+
+function p = do_win_fft(state, x)
+
+	% Power spectrum
+	s = fft(state.win .* x);
+	s_half = s(1:state.half_fft_size);
+	p = s_half .* conj(s_half);
+
+end
+
+% Wikipedia https://en.wikipedia.org/wiki/Mel_scale
+function mel = hz_to_mel(hz)
+
+	mel = 1127 * log(1 + hz / 700);
+
+end
+
+function hz = mel_to_hz(mel)
```

Comment:
I think hz = 700 * (10.^ (mel/2595) - 1);

Reply:
This is the same case. Which function in SOF is less operations exp() or 10^x?